doc/smart-switch/smart-switch-database-architecture/smart-switch-database-design.md

```diff
@@ -0,0 +1,262 @@
+
+### Architecture Design
+
+![smart-switch-database-architecture](smart-switch-database-architecture.png)
+
+#### Database services
+
+In this section, the focus is on illustrating the maintenance of DPU overlay databases within the NPU. It's essential to note that the traditional database services of both NPU and DPU remain unchanged and do not necessitate further design modifications.
+
+The management of DPU overlay databases within the NPU is orchestrated through existing SONiC database services. The daemon, named "featured," retains the responsibility for initiating, terminating, enabling, and disabling the DPU overlay database services. This interaction is facilitated using the systemctl tool.
+
+To determine the DPU number, the "featured" daemon should leverage the platform API. However, for the sake of implementation simplicity, the DPU number is extracted directly from the platform_env.conf file firstly.
+
+``` shell
+cat /usr/share/sonic/device/$PLATFORM/platform_env.conf
+NUM_DPU=2
+```
+
+To align with the established multi-ASIC design in SONiC, a new field, `"has_per_dpu_scope": "True"``, is introduced in the database feature table within config_db.json. This field plays a crucial role in ensuring that each DPU database instance is initiated within a dedicated database container. This design approach maintains consistency with SONiC's existing architecture while accommodating the specific requirements of DPU overlay databases.
+
+``` json
+# config_db.json
+
+"database": {
+    "auto_restart": "always_enabled",
+    "delayed": "False",
+    "has_global_scope": "True",
+    "has_per_asic_scope": "True",
+    "has_per_dpu_scope": "True", # New field for DPU database service
+    "high_mem_alert": "disabled",
+    "state": "always_enabled",
+    "support_syslog_rate_limit": "true"
+},
+```
+
+Within the NPU, the management of DPU overlay databases involves specific configurations. Each DPU overlay database instance is bound to the IP address of the midplane bridge (169.254.200.254 by default). The TCP port assignment follows a predictable pattern, with each DPU ID associated with a unique port (6380 + DPU ID). Additionally, the Unix domain socket path is structured as /var/run/redisdpu{DPU_ID}.
+
+Here is an example includes two DPU:
+
+``` json
+# DPU0: /var/run/redisdpu2/sonic-db/database_config.json
+"redis": {
+    "hostname": "169.254.200.254",
+    "port": 6381,
+    "unix_socket_path": "/var/run/redisdpu0/redis.sock",
+    "persistence_for_warm_boot": "yes",
+    "database_type": "dpudb"
+}
+#DPU1: /var/run/redisdpu1/sonic-db/database_config.json
+"redis": {
+    "hostname": "169.254.200.254",
+    "port": 6382,
+    "unix_socket_path": "/var/run/redisdpu1/redis.sock",
+    "persistence_for_warm_boot": "yes",
+    "database_type": "dpudb"
+}
+```
+
+There are four new tables introduction for the DPU overlay database:
+
+``` json
+"DPU_APPL_DB": {
```

Comment:
@prsunny , do you mind to help double confirm the DB name change? I think Ze is proposing modifying all DASH_* dbs to DPU_* dbs. Will any extra change needed in swss?

Also, if this is expected, we will also need to update all dash docs to make us align too.

Reply:
thanks @r12f for tagging. Yes this should be fine as it is new DB instances. Objects in the DB shall still be with DASH_* prefixes.

Reply:
on a second thought, i'm thinking why to define a with prefix DPU_*? Since this is a separate container for a specific DPU, say DPU0, why not we just keep it APPL_DB?